File: 2024/7b.cc
#include <vector>
#include <fstream>
#include <iostream>
#include <string>
#include <algorithm>
#include <cctype>

using std::ifstream;
using std::vector;
using std::string;
using std::cout;
using std::to_string;

typedef long long ll;

bool calculate(ll, ll, int, int, vector<int>&, bool);
vector<string> split(const string&, const string&);

int main(void) {

    ifstream in("7.txt");
    string line;
    vector<vector<string>> lines;
    vector<ll> sums;
    ll finalSum = 0;

    while(getline(in, line)) {
        vector<string> tokens = split(line, ": ");
        sums.push_back(stoll(tokens[0]));
        string temp = tokens[1];
        tokens = split(temp, " ");
        lines.push_back(tokens);
    }

    for(int i = 0; i < sums.size(); i++) {
        vector<int> nums;
        std::transform(lines[i].begin(), lines[i].end(), std::back_inserter(nums), [](const string& str) {
            return std::stoi(str);
        });
        if(calculate(sums[i], nums[0], 1, nums.size(), nums, true)) {
            finalSum += sums[i];
        }
    }

    cout << finalSum << "\n";

    return 0;
}

bool calculate(ll sum, ll currentSum, int index, int size, vector<int>& nums, bool performConcat) {
    if(currentSum == sum && index == size) {
        return true;
    }

    if(currentSum > sum || index == size || (currentSum == sum && index >= size)) {
        return false;
    }

    return 
            calculate(sum, currentSum + nums[index], index + 1, size, nums, performConcat) || 
            calculate(sum, currentSum * nums[index], index + 1, size, nums, performConcat) ||
            (performConcat ? 
                calculate(sum, stoll(to_string(currentSum) + to_string(nums[index])), index + 1, size, nums, performConcat) : false);
}

vector<string> split(const string& str, const string& delimiter) {
    size_t pos = 0, lastPos = 0;
    vector<string> tokens;
    while ((pos = str.find(delimiter, lastPos)) != string::npos) {
        tokens.push_back(str.substr(lastPos, pos - lastPos));
        lastPos = pos + delimiter.size();
    }
    tokens.push_back(str.substr(lastPos));
    return tokens;
}

File: 2024/7b.py
with open('7.txt') as f:
    lines = [line.strip().split(": ") for line in f]

def calculate(required_sum, current_sum, current_index, length, nums, allow_concat):

    if current_sum == required_sum and current_index == length:
        return True

    if current_index == length:
        return False

    num = int(nums[current_index])

    return (
        calculate(required_sum, current_sum + num, current_index + 1, length, nums, allow_concat) or
        calculate(required_sum, current_sum * num, current_index + 1, length, nums, allow_concat) or
        (allow_concat and calculate(required_sum, int(str(current_sum) + nums[current_index]), current_index + 1, length, nums, allow_concat))
    )

value = 0

for line in lines:

    nums = line[1].split(" ")
    length = len(nums)

    if calculate(int(line[0]), int(nums[0]), 1, length, nums, False):
        value += int(line[0])

print(value)
